// nntp/lib/nntp.js
var Q = require("q");
var net = require('net');
var sys = require('sys');
var events = require('events');

var NNTP = function(server, port) {
    this.server = server;
    this.port = port || 119;
    console.log("Creating NNTP client for '" + this.server + ":" + this.port + "'");
};
sys.inherits(NNTP, events.EventEmitter);

NNTP.prototype = {
    start: function() {
        var dfd = Q.defer();
        var client = this;
        this.connection = net.createConnection(this.port, this.server);
        this.connection.addListener("connect", function () {
            console.log("Successfully connected!");
            client.connected = true;
            dfd.resolve(true);
        });
        this.connection.addListener("error", function (error) {
            console.log("Connection failed");
            dfd.reject(error);
        });
        return dfd.promise;
    },
    authorize: function(username, password) {
        var dfd = Q.defer();
        process.nextTick(function() {dfd.resolve(true);});
        console.log("Attempting to authorize using '" + username + " / " + password + "'");
        return dfd.promise;
    },
    chooseGroup: function(group) {
        var dfd = Q.defer();
        process.nextTick(function() {dfd.resolve(true);});
        console.log("Choosing group '" + group + "'");
        return dfd.promise;
    },
    postMessage: function(title, from, body) {
        var dfd = Q.defer();
        process.nextTick(function() {dfd.resolve(true);});
        console.log("Posting message from '" + from + "' with title '" + title + "':\n====================");
        console.log(body);
        console.log("====================");
        return dfd.promise;
    },
    quit: function() {
        var dfd = Q.defer();
        process.nextTick(function() {dfd.resolve(true);});
        console.log("Quitting NNTP");
        return dfd.promise;
    }
};

module.exports = NNTP;
